### PR TITLE
fix(components): [input-number] enhanced precision

### LINKED ISSUE #7769: [Bug Report] [Component] [input-number] el-input-number数值错误问题
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.33`
- Element Plus Version: `2.2.0`
- Browser / OS: ` Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.67 Safari/537.36`
- Build Tool: `Vue CLI`

### Related Component

- `el-input-number`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuXG5jb25zdCBtc2cgPSByZWYoMSlcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxoMT57eyBtc2cgfX08L2gxPlxuICA8ZWwtaW5wdXQtbnVtYmVyIHYtbW9kZWw9XCJtc2dcIiA6cHJlY2lzaW9uPVwiMlwiIC8+XG5cbiAgPHA+XG4gICAgPGVsLWljb24gY29sb3I9XCJ2YXIoLS1lbC1jb2xvci1wcmltYXJ5KVwiPjxFbGVtZW50UGx1cyAvPjwvZWwtaWNvbj5cbiAgICBFbGVtZW50IFBsdXMge3sgRXBWZXJzaW9uIH19ICsgVnVlIHt7IHZ1ZVZlcnNpb24gfX1cbiAgPC9wPlxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydF9tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHt9XG59IiwiX28iOnt9fQ==)

### Steps to reproduce

键盘输入1.09，然后失去焦点

### What is Expected?

1.09

### What is actually happening?

1.90

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by chenxch**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

fix #7769

### CODE REVIEW

File: packages/components/input-number/src/input-number.vue [hunk 1]
```diff
@@ -167,8 +167,11 @@ export default defineComponent({
       const digits = num.toString().split('.')
       if (digits.length > 1) {
         const integer = digits[0]
-        const decimal = Math.round(+digits[1] / 10 ** (digits[1].length - pre))
-        return Number.parseFloat(`${integer}.${decimal}`)
+        const decimal = `${Math.round(
+          +`10${digits[1]}` / 10 ** (digits[1].length - pre)
```
Comment (JeremyWuuuuu) [inline packages/components/input-number/src/input-number.vue:171]: **Suggestion:**

```suggestion
          Number(`10${digits[1]}`) / 10 ** (digits[1].length - pre)
```

File: packages/components/input-number/src/input-number.vue [hunk 2]
```diff
@@ -167,8 +167,11 @@ export default defineComponent({
...
+        )}`
+        const carry = +decimal.slice(0, 2) - 10
+        return Number.parseFloat(`${+integer + carry}.${decimal.slice(2)}`)
```
Comment (JeremyWuuuuu) [inline packages/components/input-number/src/input-number.vue:174]: **Suggestion:**

```suggestion
        return Number.parseFloat(`${Number(integer) + carry}.${decimal.slice(2)}`)
```

**Review by jw-foss — COMMENTED** [review]
I would suggest that we use explicit conversion than implicit conversion in this case 

`+string` v.s `Number(string)`

Let me know what you think 🏓

**Review by btea — COMMENTED** [review]
Maybe we can adjust it directly like this:
```js
return Number.parseFloat(`${Math.round(num * 10 ** (pre + 1) / 10) / 10 ** pre}`)
```

**Review by btea — APPROVED** [review]

**Comment by qmhc** [comment]
```ts
Number.parseFloat(`${Math.round(num * 10 ** (pre + 1) / 10) / 10 ** pre}`)
```

This method also seems to be some problems, when inputs `1.3335` an fixed to 3 decimal:

image

Maybe using string schema to process is the most robust method.

**Comment by btea** [comment]
> ```ts
> Number.parseFloat(`${Math.round(num * 10 ** (pre + 1) / 10) / 10 ** pre}`)
> ```
> 
> This method also seems to be some problems, when inputs `1.3335` an fixed to 3 decimal:
> 
> image
> 
> Maybe using string schema to process is the most robust method.

what you said makes sense.

**Comment by qmhc** [comment]
I try to split it to single number to process, current cases look effective (it is written in js):

```js
function toFixed(num, pre) {
  if (pre === 0) return Math.round(num)

  const snum = String(num)
  const pointPos = snum.indexOf('.')

  if (pointPos === -1) return num

  const nums = snum.replace('.', '').split('')
  const datum = nums[pointPos + pre]

  if (!datum) return num

  let pnum

  if (datum < 5) {
    pnum = nums.slice(0, pointPos + pre).join('')
  } else {
    const pnums = nums.slice(0, pointPos + pre).map(Number)
    pnums[pointPos + pre - 1] += 1

    for (let i = pointPos + pre - 1; i >= 0; --i) {
      if (pnums[i] > 9 && i > 0) {
        pnums[i - 1] += 1
        pnums[i] -= 10
      }
    }

    pnum = pnums.join('')
  }

  return parseFloat(`${pnum.substring(0, pointPos)}.${pnum.substring(pointPos)}`)
}
```

image

**Comment by qmhc** [comment]
It seems like a more concise method:

```js
function toFixed(num, pre) {
  if (pre === 0) return Math.round(num)

  let snum = String(num)
  const pointPos = snum.indexOf('.')

  if (pointPos === -1) return num

  const nums = snum.replace('.', '').split('')
  const datum = nums[pointPos + pre]

  if (!datum) return num

  const length = snum.length

  if (snum.charAt(length - 1) === '5') {
    snum = snum.substring(0, length - 1) + '6'
  }

  return parseFloat(Number(snum).toFixed(pre))
}
```

image

**Comment by chenxch** [comment]
@qmhc First of all, thank you very much for your many suggestions on this part. I will replace `5` with `6` to see if it is safe. If there is no problem, I will adopt your solution, and I will also provide my original processing method.
```ts
const toPrecision = (num: number, pre?: number) => {
      if (isUndefined(pre)) pre = numPrecision.value
      const digits = num.toString().split('.')
      if (digits.length > 1) {
        const integer = digits[0]
        const decimal = `${Math.round(
          Number(`10${digits[1]}`) / 10 ** (digits[1].length - pre)
        )}`
        const carry = +decimal.slice(0, 2) - 10
        return Number.parseFloat(
          `${Number(integer) + carry}.${decimal.slice(2)}`
        )
      }
      return Number.parseFloat(`${Math.round(num * 10 ** pre) / 10 ** pre}`)
    }
```

**Comment by chenxch** [comment]
@qmhc Well done, use your scheme.

**Comment by chenxch** [comment]
@btea Need your confirmation again.

**Review by btea — COMMENTED** [review]
LGTM 👍

**Pull request merged**